Proxy performance

tw4452852 opened the pull request:
update issue #939

mholt commented:
@tw4452852 I'm always super glad to see your pull requests. :smile: I'm excited to look this over in more detail tomorrow (after I've had some sleep)!

mholt commented on caddyhttp/httpserver/replacer.go:
```diff
@@ -37,8 +37,8 @@ type Replacer interface {
-	replacements       map[string]string
-	customReplacements map[string]string
+	replacements       map[string]func() string
+	customReplacements map[string]func() string
```
Ahhh... this makes sense. Does using functions to defer evaluation of these values have anything to do with #916 (where the `Host: {host}` was 'cached' between different hostnames)?

tw4452852 commented:
@mholt I don't think this change has something to do with #916. This change just defer the preparation of `replacer`. Here is the excerpt of profile of `NewReplacer` which is created on every request.

```    
  10ms      450ms (flat, cum) 10.61% of Total
         .          .     52:// of empty string (can still be empty string).
         .          .     53:func NewReplacer(r *http.Request, rr *ResponseRecorder, emptyValue string) Replacer {
         .          .     54:   rep := &replacer{
         .          .     55:           responseRecorder:   rr,
         .          .     56:           customReplacements: make(map[string]string),
         .       10ms     57:           replacements: map[string]string{
         .       20ms     58:                   "{method}": r.Method,
         .          .     59:                   "{scheme}": func() string {
         .          .     60:                           if r.TLS != nil {
         .          .     61:                                   return "https"
         .          .     62:                           }
         .          .     63:                           return "http"
         .          .     64:                   }(),
         .          .     65:                   "{hostname}": func() string {
         .          .     66:                           name, err := os.Hostname()
         .          .     67:                           if err != nil {
         .          .     68:                                   return ""
         .          .     69:                           }
         .          .     70:                           return name
         .      190ms     71:                   }(),
         .          .     72:                   "{host}": r.Host,
         .          .     73:                   "{hostonly}": func() string {
         .          .     74:                           host, _, err := net.SplitHostPort(r.Host)
         .          .     75:                           if err != nil {
         .          .     76:                                   return r.Host
         .          .     77:                           }
         .          .     78:                           return host
         .          .     79:                   }(),
         .          .     80:                   "{path}":          r.URL.Path,
         .       20ms     81:                   "{path_escaped}":  url.QueryEscape(r.URL.Path),
         .       10ms     82:                   "{query}":         r.URL.RawQuery,
         .       40ms     83:                   "{query_escaped}": url.QueryEscape(r.URL.RawQuery),
         .          .     84:                   "{fragment}":      r.URL.Fragment,
         .       10ms     85:                   "{proto}":         r.Proto,
         .       10ms     86:                   "{remote}": func() string {
         .          .     87:                           if fwdFor := r.Header.Get("X-Forwarded-For"); fwdFor != "" {
         .          .     88:                                   return fwdFor
         .          .     89:                           }
         .          .     90:                           host, _, err := net.SplitHostPort(r.RemoteAddr)
         .          .     91:                           if err != nil {
         .          .     92:                                   return r.RemoteAddr
         .          .     93:                           }
         .          .     94:                           return host
         .       10ms     95:                   }(),
         .          .     96:                   "{port}": func() string {
         .          .     97:                           _, port, err := net.SplitHostPort(r.RemoteAddr)
         .          .     98:                           if err != nil {
         .          .     99:                                   return ""
         .          .    100:                           }
         .          .    101:                           return port
         .          .    102:                   }(),
         .       30ms    103:                   "{uri}":         r.URL.RequestURI(),
         .       10ms    104:                   "{uri_escaped}": url.QueryEscape(r.URL.RequestURI()),
         .       30ms    105:                   "{when}":        time.Now().Format(timeFormat),
         .          .    106:                   "{file}": func() string {
         .          .    107:                           _, file := path.Split(r.URL.Path)
         .          .    108:                           return file
         .          .    109:                   }(),
         .          .    110:                   "{dir}": func() string {
         .          .    111:                           dir, _ := path.Split(r.URL.Path)
         .          .    112:                           return dir
         .          .    113:                   }(),
         .          .    114:                   "{request}": func() string {
         .          .    115:                           dump, err := httputil.DumpRequest(r, false)
         .          .    116:                           if err != nil {
         .          .    117:                                   return ""
         .          .    118:                           }
         .          .    119:
         .          .    120:                           return requestReplacer.Replace(string(dump))
      10ms       50ms    121:                   }(),
         .          .    122:           },
         .       10ms    123:           emptyValue: emptyValue,
         .          .    124:   }

```

We could see a lot of time spent on the preparation of replacements.
And in most case, we only use some part of these replacements,
So I just work out this change to defer these preparation until it is used actually.

Now the profile is looks much better:

```
      10ms      200ms (flat, cum)  5.10% of Total                                                                                                                                            [33/13709]
         .          .     51:// emptyValue should be the string that is used in place
         .          .     52:// of empty string (can still be empty string).
         .          .     53:func NewReplacer(r *http.Request, rr *ResponseRecorder, emptyValue string) Replacer {
         .          .     54:   rep := &replacer{
         .          .     55:           responseRecorder:   rr,
         .       10ms     56:           customReplacements: make(map[string]func() string),
         .          .     57:           replacements: map[string]func() string{
         .       20ms     58:                   "{method}": func() string { return r.Method },
         .          .     59:                   "{scheme}": func() string {
         .          .     60:                           if r.TLS != nil {
         .          .     61:                                   return "https"
         .          .     62:                           }
         .          .     63:                           return "http"
         .          .     64:                   },
         .          .     65:                   "{hostname}": func() string {
         .          .     66:                           name, err := os.Hostname()
         .          .     67:                           if err != nil {
         .          .     68:                                   return ""
         .          .     69:                           }
         .          .     70:                           return name
         .          .     71:                   },
         .          .     72:                   "{host}": func() string { return r.Host },
         .          .     73:                   "{hostonly}": func() string {
         .          .     74:                           host, _, err := net.SplitHostPort(r.Host)
         .          .     75:                           if err != nil {
         .          .     76:                                   return r.Host
         .          .     77:                           }
         .          .     78:                           return host
         .          .     79:                   },
         .          .     80:                   "{path}":          func() string { return r.URL.Path },
         .          .     81:                   "{path_escaped}":  func() string { return url.QueryEscape(r.URL.Path) },
         .          .     82:                   "{query}":         func() string { return r.URL.RawQuery },
         .       30ms     83:                   "{query_escaped}": func() string { return url.QueryEscape(r.URL.RawQuery) },
         .       20ms     84:                   "{fragment}":      func() string { return r.URL.Fragment },
         .       20ms     85:                   "{proto}":         func() string { return r.Proto },
         .       10ms     86:                   "{remote}": func() string {
         .          .     87:                           if fwdFor := r.Header.Get("X-Forwarded-For"); fwdFor != "" {
         .          .     88:                                   return fwdFor
         .          .     89:                           }
         .          .     90:                           host, _, err := net.SplitHostPort(r.RemoteAddr)
         .          .     91:                           if err != nil {
         .          .     92:                                   return r.RemoteAddr
         .          .     93:                           }
         .          .     94:                           return host
         .          .     95:                   },
         .       10ms     96:                   "{port}": func() string {
         .          .     97:                           _, port, err := net.SplitHostPort(r.RemoteAddr)
         .          .     98:                           if err != nil {
         .          .     99:                                   return ""
         .          .    100:                           }
         .          .    101:                           return port
         .          .    102:                   },
         .       50ms    103:                   "{uri}":         func() string { return r.URL.RequestURI() },
         .          .    104:                   "{uri_escaped}": func() string { return url.QueryEscape(r.URL.RequestURI()) },
         .          .    105:                   "{when}":        func() string { return time.Now().Format(timeFormat) },
         .          .    106:                   "{file}": func() string {
         .          .    107:                           _, file := path.Split(r.URL.Path)
         .          .    108:                           return file
         .          .    109:                   },
         .          .    110:                   "{dir}": func() string {
         .          .    111:                           dir, _ := path.Split(r.URL.Path)
         .          .    112:                           return dir
         .          .    113:                   },
         .       10ms    114:                   "{request}": func() string {
         .          .    115:                           dump, err := httputil.DumpRequest(r, false)
         .          .    116:                           if err != nil {
         .          .    117:                                   return ""
         .          .    118:                           }
         .          .    119:
         .          .    120:                           return requestReplacer.Replace(string(dump))
         .          .    121:                   },
         .          .    122:           },
      10ms       10ms    123:           emptyValue: emptyValue,
         .          .    124:   }
```

mholt commented on caddyhttp/proxy/reverseproxy.go:
```diff
@@ -214,6 +220,9 @@ func (p *ReverseProxy) ServeHTTP(rw http.ResponseWriter, outreq *http.Request, r
 
 func (p *ReverseProxy) copyResponse(dst io.Writer, src io.Reader) {
+	buf := bufferPool.Get()
+	defer bufferPool.Put(buf)
```
Nice - do you happen to know what the approximate/general improvement is using this buffer pool?

tw4452852 replied:
@mholt I find this during the memory profiling, here it is:

```
       4MB   917.01MB (flat, cum) 81.12% of Total
         .          .    217:   if p.FlushInterval != 0 {
         .          .    218:           if wf, ok := dst.(writeFlusher); ok {
         .          .    219:                   mlw := &maxLatencyWriter{
         .          .    220:                           dst:     wf,
         .          .    221:                           latency: p.FlushInterval,
       4MB        4MB    222:                           done:    make(chan bool),
         .          .    223:                   }
         .          .    224:                   go mlw.flushLoop()
         .          .    225:                   defer mlw.stop()
         .          .    226:                   dst = mlw
         .          .    227:           }
         .          .    228:   }
         .   913.01MB    229:   io.Copy(dst, src)
         .          .    230:}
```

As you see `io.Copy` in `proxy.(*ReverseProxy).copyResponse` allocated a lot of memory and the culprit is:

```
 912.43MB   913.01MB (flat, cum) 80.77% of Total
         .          .    382:   // Similarly, if the writer has a ReadFrom method, use it to do the copy.
         .          .    383:   if rt, ok := dst.(ReaderFrom); ok {
         .          .    384:           return rt.ReadFrom(src)
         .          .    385:   }
         .          .    386:   if buf == nil {
  912.43MB   912.43MB    387:           buf = make([]byte, 32*1024)
         .          .    388:   }
         .          .    389:   for {
         .          .    390:           nr, er := src.Read(buf)
         .          .    391:           if nr > 0 {
         .   596.16kB    392:                   nw, ew := dst.Write(buf[0:nr])
         .          .    393:                   if nw > 0 {
         .          .    394:                           written += int64(nw)
         .          .    395:                   }
         .          .    396:                   if ew != nil {
         .          .    397:                           err = ew
```

It allocate a 32k buffer for every copying(in other word, every request) because we did not supply the buffer, so I  think we could use buffer pool to offer and recycle the temporary buffer.
And it seems works well:

```
ROUTINE ======================== github.com/mholt/caddy/caddyhttp/proxy.(*ReverseProxy).copyResponse in /home/tw/golib/src/github.com/mholt/caddy/caddyhttp/proxy/reverseproxy.go
       4MB    19.05MB (flat, cum)  8.37% of Total
         .          .    218:
         .          .    219:   return nil
         .          .    220:}
         .          .    221:
         .          .    222:func (p *ReverseProxy) copyResponse(dst io.Writer, src io.Reader) {
         .    13.93MB    223:   buf := bufferPool.Get()
         .          .    224:   defer bufferPool.Put(buf)
         .          .    225:
         .          .    226:   if p.FlushInterval != 0 {
         .          .    227:           if wf, ok := dst.(writeFlusher); ok {
         .          .    228:                   mlw := &maxLatencyWriter{
         .          .    229:                           dst:     wf,
         .          .    230:                           latency: p.FlushInterval,
       4MB        4MB    231:                           done:    make(chan bool),
         .          .    232:                   }
         .          .    233:                   go mlw.flushLoop()
         .          .    234:                   defer mlw.stop()
         .          .    235:                   dst = mlw
         .          .    236:           }
         .          .    237:   }
         .     1.12MB    238:   io.CopyBuffer(dst, src, buf.([]byte))
         .          .    239:}
         .          .    240:
         .          .    241:type writeFlusher interface {
         .          .    242:   io.Writer
         .          .    243:   http.Flusher

```

mholt commented:
This is awesome, thanks @tw4452852! It definitely looks like it will help speed up the proxy a little bit and save a lot of memory use under heavy load.

abiosoft commented:
This is awesome Thanks @tw4452852.
